File: 백준/Gold/1987. 알파벳/알파벳.py
r, c = map(int,input().split())
visit = [False]*26
board = [input() for _ in range(r)]
answer = 0
def dfs(x,y,l):
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0<=nx<r and 0<=ny<c:
            if visit[ord(board[nx][ny])-65]:
                global answer
                answer = max(answer,l)
            else:
                visit[ord(board[nx][ny])-65] = True
                dfs(nx,ny,l+1)
                visit[ord(board[nx][ny])-65] = False
visit[ord(board[0][0])-65]=True
dfs(0,0,1)
print(answer)

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int R, C, answer = 0;
	static char[][] map;
	static final int[][] delta = {{-1,0},{1,0},{0,-1},{0,1}};
	static long[] check = new long[26];
	public static void main(String[] args) throws Exception {
		st = new StringTokenizer(br.readLine());
		R = Integer.parseInt(st.nextToken()); C = Integer.parseInt(st.nextToken());
		map = new char[R][C];
		long l = 1L;
		for(int i=0;i<26;i++,l<<=1)check[i] = l;
		for(int i=0;i<R;i++) map[i] = br.readLine().toCharArray();
		recur(0,0,1,1L<<(map[0][0]-'A'));
		System.out.print(answer);
	}
	static void recur(int r, int c, int l, long visit){
		for(int i=0;i<4;i++){
			int nr = r + delta[i][0], nc = c + delta[i][1];
			if(isIn(nr,nc)){
				long check = 1L<<(map[nr][nc]-'A');
				if((visit&check)!=0L) answer = Math.max(answer,l);
				else{
					recur(nr,nc,l+1,visit|check);
				}
			}
		}
	}
	static boolean isIn(int nr, int nc){
		return 0<=nr&&nr<R&&0<=nc&&nc<C;
	}
}